feat: bindings in `or`-patterns (#3807)
dfinity/motoko: doc/md/language-manual.md

The result of matching `<pat1> or <pat2>` against a value is the result of matching `<pat1>`, if it succeeds, or the result of matching `<pat2>`, if the first match fails.

> (Note, a disjunctive pattern may contain identifier (`<id>`) patterns with the restriction that both alternatives must bind the same identifiers and these must

Reviewer: ```suggestion
(Note, a disjunctive pattern may contain identifier (`<id>`) patterns with the restriction that both alternatives must bind the same set of identifiers with equivalent types.
```

> (Note, a disjunctive pattern may contain identifier (`<id>`) patterns with the restriction that both alternatives must bind the same identifiers and these must
> obtain values of the same type.)

Reviewer: ```suggestion
(Note, a disjunctive pattern may contain identifier (`<id>`) patterns with the restriction that both alternatives must bind the same set of identifiers. Each identifier's type is the least upper bound of its type in `<pat1>` and `<pat2>`.
```
Ahh, but I now see that the source typing rules require equality but the ir ones just use the lub....

Author: Yes, the type checker wants equality. The error message should make that clear. Further downstream I am using `lub` for that being the correct connective if we want to water up the type-checker at a later point in time. It does the right thing when the types are equal.

I chose strict equality, so that no information is being lost when folks use `or` patterns. Otherwise we could get into the situation where the `lub` becomes `Any` and we'd have to warn like for the array construction case. If the user wants to deal with different types, that is still totally possible by writing out both `case`s.

Author: The other reason to use `lub` in the IR is that `typing.ml` assigns type `Non` to dead branch patten bindings.

Reviewer 2: In a language with subtyping not using a lub here would be fishy, because that immediately breaks subtype weakening. Changing the type of some variable to a subtype should not affect typeability of a program -- that would be an obvious "refactoring hazard". Of course, warning if the lub is lossy is still a good idea.

Reviewer: It looks to me like you are taking the lub now, after warning (in typing.ml). Looks good.
